src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseSpanBasedStringConcat.cs

```diff
@@ -0,0 +1,310 @@
+                {
+                    builder.Add(current.RightOperand);
+                }
+                else
+                {
+                    GoLeft(rightBinary);
+                }
+            }
+
+            return builder.ToImmutable();
+
+            void GoLeft(IBinaryOperation operation)
+            {
+                IBinaryOperation? current = operation;
+                while (current is not null && current.OperatorKind == root.OperatorKind)
+                {
+                    stack.Push(current);
+                    current = current.LeftOperand as IBinaryOperation;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Remove the built in implicit conversion on operands to concat.
+        /// In VB, the conversion can be to either string or object.
+        /// In C#, the conversion is always to object.
+        /// </summary>
+        internal static IOperation WalkDownBuiltInImplicitConversionOnConcatOperand(IOperation operand)
+        {
+            if (operand is not IConversionOperation conversion)
+                return operand;
+            if (!conversion.IsImplicit || conversion.Conversion.IsUserDefined)
+                return conversion;
+
+            switch (conversion.Language)
+            {
+                case LanguageNames.CSharp when conversion.Type.SpecialType is SpecialType.System_Object:
+                case LanguageNames.VisualBasic when conversion.Type.SpecialType is SpecialType.System_Object or SpecialType.System_String:
+                    return conversion.Operand;
+                default:
+                    return conversion;
+            }
+        }
+
+        // Use readonly struct instead of record type to save on allocations, since it's not passed by-value.
+        // We aren't comparing these.
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+        internal readonly struct RequiredSymbols
+#pragma warning restore CA1815 // Override equals and operator equals on value types
+        {
+            private RequiredSymbols(
+                INamedTypeSymbol stringType, INamedTypeSymbol roscharType,
+                IMethodSymbol substring1, IMethodSymbol substring2,
+                IMethodSymbol asSpan1, IMethodSymbol asSpan2)
+            {
+                StringType = stringType;
+                ReadOnlySpanOfCharType = roscharType;
+                SubstringStart = substring1;
+                SubstringStartLength = substring2;
+                AsSpanStart = asSpan1;
+                AsSpanStartLength = asSpan2;
+
+                RoslynDebug.Assert(
+                    StringType is not null && ReadOnlySpanOfCharType is not null &&
+                    SubstringStart is not null && SubstringStartLength is not null &&
+                    AsSpanStart is not null && AsSpanStartLength is not null);
```

Comment:
I don't believe this is required - the repo is nullable enabled and defining parameters without `?` ensures that a potential value passed in at callsite is flagged by the compiler with nullable warning.
```suggestion
```